File: Difficulty/Hard/[99] Recover Binary Search Tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def recoverTree(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        s, node1, node2, pre = [], None, None, None
        while root != None or s:
            while root != None:
                s.append(root)
                root = root.left
            root = s.pop()
            if pre != None and pre.val > root.val:
                if node1 == None:
                    node1 = pre
                    node2 = root
                else:
                    node2 = root
            pre = root
            root = root.right
        if node1 is None and node2 is None:
            node1.val, node2.val = node2.val, node1.val

File: Difficulty/Hard/[99] Recover Binary Search Tree/solution.cpp
/*
 * @lc app=leetcode id=99 lang=cpp
 *
 * [99] Recover Binary Search Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <stack>

class Solution {
public:
    void recoverTree(TreeNode* root) {
        stack<TreeNode*> s;
        TreeNode* node1 = nullptr, *node2 = nullptr, *pre = nullptr;
        while (root != nullptr || !s.empty()) {
            while(root != nullptr){
                s.push(root);
                root = root->left;
            }
            root = s.top();
            s.pop();
            if (pre != nullptr && pre->val > root->val) {
                if (node1 == nullptr)
                    {node1 = pre; node2 = root;}
                else{
                     node2 = root; 
                     swap(node1->val, node2->val);
                     return;
                }
            }
            pre = root;
            root = root->right;
        }
        if (node1 != nullptr && node2 != nullptr)
            swap(node1->val, node2->val);
    }
};
// @lc code=end

